subsys/bluetooth/host/mesh/cfg_srv.c

```diff
@@ -1128,6 +1128,54 @@ static u8_t va_add(u8_t *label_uuid, u16_t *addr)
 	return STATUS_SUCCESS;
 }
 
+static u8_t va_del(u8_t *label_uuid, u16_t *addr)
+{
+	int i;
+
+	if (addr) {
+		*addr = BT_MESH_ADDR_UNASSIGNED;
+	}
+
+	for (i = 0; i < ARRAY_SIZE(labels); i++) {
+		if (!memcmp(labels[i].uuid, label_uuid, 16)) {
+			if (addr) {
+				*addr = labels[i].addr;
+			}
+			memset(&labels[i], 0, sizeof(labels[i]));
+
+			return STATUS_SUCCESS;
+		}
+	}
+
+	return STATUS_CANNOT_REMOVE;
+}
+
+static void mod_sub_clear(struct bt_mesh_model *mod)
+{
+	struct label *label;
+	u8_t *label_uuid;
+	int i;
+
+	/* Clear stored labels related to this model */
+	for (i = 0; i < ARRAY_SIZE(mod->groups); i++) {
+		if (!BT_MESH_ADDR_IS_VIRTUAL(mod->groups[i])) {
+			continue;
+		}
+
+		label_uuid = bt_mesh_label_uuid_get(mod->groups[i]);
+		if (!label_uuid) {
+			BT_ERR("Label UUID not found");
+			continue;
+		}
+
+		label = CONTAINER_OF(label_uuid, struct label, uuid);
+		memset(label, 0, sizeof(*label));
```

Comment:
It looks to me like this is broken in case another model subscribes to the same virtual label UUID. I think that's the reason why I never had code to remove them. If this really needs fixing there are two ways I can think of: either add refcounting to the labels, or then iterate all models looking for any other users before deciding to remove a label for good.

Reply:
Indeed, you have right. I like the refcounting idea